[DOTORG-136]On the client, admin should be able to trigger campaign status mutations

nguyenaiden opened the pull request:
### Description
* Expand admin_campaigns button objects to handle status mutations
* Expand list button group to have a `disabled` state (***completed*** status can not be further mutated)
* Add campaign action to handle campaign status update
* Add componentClickHandler to campaign_list component to handle status update
* Update container to map updateCampaignStatus action to the props

### Testing plan
* Manually test front-end/controller interaction on view and verify in DB
* Add Action test for updateCampaignStatus

Commit:
Expand admin_campaigns button objects to handle status mutations
Expand list button group to accept disabled status
Add campaign action to handle campaign status update
Add componentClickHandler to campaign_list component to handle status update
Update container to map updateCampaignStatus action to the props

nguyenaiden committed:
Merge branch 'master' into campaign-status-transition-view

Commit:
Refactored for statusToggle functions to need 1 invocation

Commit:
Merge branch 'campaign-status-transition-view' of https://github.com/Twilio-org/phonebank into campaign-status-transition-view

nguyenaiden committed:
Merge branch 'master' into campaign-status-transition-view

nguyenaiden committed:
Merge branch 'master' into campaign-status-transition-view

Commit:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into campaign-status-transition-view

Commit:
Connected front end buttons with back-end routes from PR #156

Commit:
Merge branch 'campaign-status-transition-view' of https://github.com/Twilio-org/phonebank into campaign-status-transition-view

nguyenaiden committed:
Merge branch 'master' into campaign-status-transition-view

Commit:
UpdateCampaignStatus action returns the campaign after update
Add test in admin_campaigns tests for PUT request

andioneto commented on public/src/actions/campaign.jsx:
```diff
@@ -105,3 +105,17 @@ export function fetchCampaignsByUser(userId, current_campaign) {
+  )
+  .then((campaign) => {
+    dispatch(fetchAll());
+    return campaign;
```
no need to return the campaign obj here....

nguyenaiden replied:
This was for testing purposes. To verify that the object coming back is correct.

perfectlynormalbeast replied:
So are we removing it then?

perfectlynormalbeast commented on public/src/components/campaign/campaign_list.jsx:
```diff
@@ -15,21 +15,15 @@ export default class CampaignList extends Component {
     const { admin_campaigns, volunteer_campaigns } = buttons_obj;
     const { redirect_path } = campaignHeaders;
 
-    function handleClick(id, historyObj, campaign) {
```
I assume these were unused in the list view for some reason?

nguyenaiden replied:
They weren't used until the user status mutations/toggle was introduced.

andioneto replied:
@nguyenaiden isn't this necessary for assigning users to campaigns in the volunteer campaign view?

nguyenaiden replied:
That function was mapped to the button group, but it was never used since the handler function for ```admin_campaigns``` in `button_objs.jsx`  didn't map to any handleClick until this PR.

andioneto commented on public/src/components/campaign/campaign_list.jsx:
```diff
@@ -15,21 +15,15 @@ export default class CampaignList extends Component {
       auth,
       history,
-      setCurrentCampaign
+      updateCampaignStatus
```
Why did you remove setCurrentCampaign? I think you're using this in your current ticket, right?

nguyenaiden replied:
I'm adding that back in for the next ticket. Removing it so it passes the linter because I'm not actually using that right now.

andioneto replied:
hmmm maybe just comment it out?

nguyenaiden replied:
Addressed per our conversation

andioneto commented on public/src/components/campaign/campaign_list.jsx:
```diff
@@ -42,7 +36,7 @@ export default class CampaignList extends Component {
                                    admin_campaigns :
                                    volunteer_campaigns}
-              componentClickHandler={handleClick}
+              componentClickHandler={updateCampaignStatus}
```
Maybe check this against your current ticket. I think I remember this was changed for the campaign stats as well... just want to make sure the changes are compatible.

nguyenaiden replied:
This componentClickHandler is used to transition statuses. For the campaign stats page, I don't need a click handler because it's just a history.push to a different page. No need for a separate function.

andioneto replied:
@nguyenaiden see comment above. as long as we didn't delete the action call that assigns users to campaigns this is fine

andioneto commented on public/src/components/common/button_objs.jsx:
```diff
@@ -1,3 +1,19 @@
+function statusToggleActivePause(status) {
+  if (status === 'draft' || status === 'pause') {
+    return ['Activate', 'active', 'success', false];
```
Maybe these should be maps/objects? would make this logic a bit cleaner to read... also potentially less buggy (someone could modify the order here and the component would break.

nguyenaiden replied:
I agree with using a map here. I'll make the appropriate changes.

perfectlynormalbeast commented on server/util/controller_helpers.js:
```diff
@@ -23,11 +23,16 @@ export function validCampaignStatusUpdate(prevStatus, currStatus) {
     return { name, title, description, contact_lists_id, script_id };
   }
-  throw new Error('Cannot edit campaign that is not in draft.');
+  const isValid = _.every(Object.keys(params), param => !!params[param] === false);
```
This line is confusing. A comment here would be helpful.

perfectlynormalbeast commented on server/util/controller_helpers.js:
```diff
@@ -23,11 +23,16 @@ export function validCampaignStatusUpdate(prevStatus, currStatus) {
+  if (!isValid) {
+    throw new Error('Cannot edit campaign that is not in draft.');
+  }
+  return true;
```
The previous return statement returns an object with update params. This return statement just returns a boolean. We need to be consistent here. How does the calling method use this value?

nguyenaiden replied:
👍

andioneto commented on server/util/controller_helpers.js:
```diff
@@ -23,11 +23,16 @@ export function validCampaignStatusUpdate(prevStatus, currStatus) {
+  const params = { name, title, description, contact_lists_id, script_id };
   if (currentStatus === 'draft') {
-    const { name, title, description, contact_lists_id, script_id } = requestBody;
     return { name, title, description, contact_lists_id, script_id };
```
why not just return the 'params' obj.... I think this is repetitive.

nguyenaiden replied:
Agreed.

andioneto reviewed (changes requested):
Looks good, a few main takeaways:
1. Maybe check to make sure you don't need setCurrentCampaign anymore, I think it would be good to keep it.
2. Maybe clean up logic in the toggle disable function (button obj file)
3. some dry-ifyng

perfectlynormalbeast commented on public/src/components/common/list_btngrp.jsx:
```diff
@@ -7,9 +7,10 @@ const ListBtnGroup = (props) => {
       {
         buttons.map((button) => {
-          const { style, text, handler, size, key } = button;
+          const { style, text, handler, size, key, disabled } = button;
```
Is this affecting any other button groups?

nguyenaiden replied:
It's not affecting the other ones in a way that they don't need the `disabled` attributes so it's silently ignored. I needed this to block button clicks for unavailable transitions depending on the current campaign status.

perfectlynormalbeast reviewed (changes requested):
A few comments

Commit:
Changed button obj logic to use a statusMap

Commit:
Changed controller helper to have consistent logic

Commit:
Commented out setCurrentCampaign for future uses

Commit:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into campaign-status-transition-view

Commit:
Refactored neccesary functions for user to join campaigns

nguyenaiden committed:
Merge branch 'master' into campaign-status-transition-view

nguyenaiden committed:
Merge branch 'master' into campaign-status-transition-view

andioneto reviewed (dismissed):
Looks good to go 👍

andioneto committed:
Server endpoint for campaign stats (#166)

splitResponses
split responses into an array of response strings
[params: responseString (string)]
checkCampaignNotDraft
goal: throw error if metrics request is made for a campaign that is still in 'draft'
[params: campaignId (int), res (response obj)]
createStatusMap
goal: create map of statuses to frequency for the campaign (initialized to 0); mapped from 'status_distribution' prop in the metrics object
[params: metricsObj (object)]
createOutcomeMap
goal: create map of outcomes to frequency for the campaign (initialized to 0); mapped from 'outcome_distribution' prop in the metrics object
[params: metricsObj (object)]
createResponseMap
goal: create map of responses to specific questions, to frequency for the campaign (initialized to 0); mapped from ' response_stats' prop in the metrics object
props are question ids, which map to another object that represents the frequency of each reponse
countStatusAndOutcomeFrequency
goal: iterate over calls and 'count' the frequencies of each outcoome/status
[params: callsArray (array of call models/ objs)]
countResponseFrequency
goal: count the frequency of responses for a specific questions within a specific campaign
[params: responsesArray(array), metricsObj(obj), questionId(int)]
first splits the serialized repsonses into sub-arrays, then performs a shallow flatten, then counts freq.
handleQuestionResponses
goal: promisifies fetching responses by question_Id and call_id in a campaign (left join that returns response, call_id, question_id

new services: responses and questions

perfectlynormalbeast reviewed (dismissed):
One tiny comment, otherwise looks good if tested as intended

perfectlynormalbeast commented:
Oh btw the controller_helpers file moved so you may need to move your method there too.

lupitadavila committed:
DOTORG-156: Calls Page Styling (#175)

* Start styling of calls page

* Complete styling and less file organization, update test

* Fix small fixes requested

andioneto committed:
Merge branch 'campaign-status-transition-view' of https://github.com/Twilio-org/phonebank into campaign-status-transition-view

andioneto reviewed (approved):
Good to go.

perfectlynormalbeast reviewed (approved):
🚀
